cdk/ec2/lib/user-data/billings-user-data.sh

```diff
@@ -0,0 +1,106 @@
+#!/bin/bash
+
+# Install necessary packages
+# retry due to the error like this: "RPM: error: can't create transaction lock on /var/lib/rpm/.rpm.lock (Resource temporarily unavailable)"
+# sometime the ec2 instances are launched too fast that when the script runs something is not really ready
+# see https://repost.aws/questions/QU_tj7NQl6ReKoG53zzEqYOw/amazon-linux-2023-issue-with-installing-packages-with-cloud-init
+max_attempts=5
+attempt_num=1
+success=false
+while [ $success = false ] && [ $attempt_num -le $max_attempts ]; do
+  echo "Trying yum install"
+  yum update -y
+  yum install -y java-17-amazon-corretto-devel git tmux wget jq
+  # Check the exit code of the command
+  if [ $? -eq 0 ]; then
+    echo "Yum install succeeded"
+    success=true
+  else
+    echo "Attempt $attempt_num failed. Sleeping for 3 seconds and trying again..."
+    sleep 3
+    ((attempt_num++))
+  fi
+done
+
+# Install CloudWatch Agent
+wget https://amazoncloudwatch-agent.s3.amazonaws.com/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm
+sudo rpm -U ./amazon-cloudwatch-agent.rpm
+
+# Create CloudWatch Agent configuration file
+cat <<'EOC' > /amazon-cloudwatch-agent.json
+{
+  "traces": {
+    "traces_collected": {
+      "application_signals": {}
+    }
+  },
+  "logs": {
+    "metrics_collected": {
+      "application_signals": {
+        "rules":[
+               {
+                  "selectors":[
+                     {
+                        "dimension":"RemoteService",
+                        "match":"169.254.169.254*"
+                     }
+                  ],
+                  "action":"drop"
+               }
+            ]
+      }
+    }
+  }
+}
+EOC
+
+# Start CloudWatch Agent
+/opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl \
+  -a fetch-config -m ec2 -s \
+  -c file:/amazon-cloudwatch-agent.json
+
+
+
+# Switch to ec2-user to run commands
+sudo -iu ec2-user bash <<'EOF'
+set -x
+# Set home directory
+cd ~
+
+# Clone the application repository
+git clone --branch cdk-setup https://github.com/pxaws/application-signals-demo.git
+cd application-signals-demo/
+
+# Function to wait for a URL to become accessible
+wait_for_url() {
+  local url=$1
+  echo "Waiting for $url to be accessible..."
+  until curl --silent --head --fail "$url"; do
+    echo "$url is not accessible. Retrying in 10 seconds..."
+    sleep 10
+  done
+  echo "$url is now accessible"
+}
+
+# Wait for config and discovery server to be ready
+wait_for_url "http://setup.demo.local:8888"
```

Comment:
The `setup.demo.local` is a private DNS name. We need to select a domain `demo` and a namespace `local` for it. I think we can use cloudwatch agent config to replace the names so that they can show up as `config-server` and `discovery-server` in service map.